File: Backend/renderer/src/main/java/io/sparkled/renderer/effect/EffectRenderer.java
package io.sparkled.renderer.effect;

import io.sparkled.model.animation.effect.Effect;
import io.sparkled.model.render.RenderedChannel;
import io.sparkled.model.render.RenderedFrame;
import io.sparkled.renderer.context.RenderContext;
import io.sparkled.renderer.easing.EasingFunction;
import io.sparkled.renderer.easing.EasingFunctions;

public abstract class EffectRenderer {

    public final void render(RenderedChannel channel, RenderedFrame frame, Effect effect) {
        float progress = getProgress(frame, effect);
        RenderContext ctx = new RenderContext(channel, frame, effect, progress);
        render(ctx);
    }

    private float getProgress(RenderedFrame frame, Effect effect) {
        EasingFunction easingFunction = EasingFunctions.get(effect.getEasing().getType());

        int startFrame = effect.getStartFrame();
        int duration = effect.getEndFrame() - startFrame + 1;

        int repetitions = effect.getRepetitions();
        int framesPerRepetition = duration / repetitions;
        int repetitionFrameNumber = (frame.getFrameNumber() - startFrame) % (framesPerRepetition + 1);


        float progress = easingFunction.getProgress(effect.getEasing(), repetitionFrameNumber, framesPerRepetition);
        if (progress < 0 || progress > 1) {
            throw new IllegalStateException("Animation progress is out of bounds: " + progress);
        }

        return effect.isReverse() ? 1 - progress : progress;
    }

    protected abstract void render(RenderContext ctx);
}

File: renderer/src/main/kotlin/io/sparkled/renderer/effect/EffectRenderer.kt
package io.sparkled.renderer.effect

import io.sparkled.model.animation.effect.Effect
import io.sparkled.model.entity.Sequence
import io.sparkled.model.entity.StageProp
import io.sparkled.model.render.RenderedFrame
import io.sparkled.model.render.RenderedStagePropData
import io.sparkled.renderer.context.RenderContext
import io.sparkled.renderer.easing.EasingFunctions

abstract class EffectRenderer {

    fun render(sequence: Sequence, channel: RenderedStagePropData, frame: RenderedFrame, stageProp: StageProp, effect: Effect) {
        val progress = getProgress(frame, effect)
        val ctx = RenderContext(sequence, channel, frame, stageProp, effect, progress)
        render(ctx)
    }

    private fun getProgress(frame: RenderedFrame, effect: Effect): Float {
        val easingFunction = EasingFunctions[effect.easing.type]

        val currentFrame = frame.frameNumber - effect.startFrame
        val startFrame = effect.startFrame
        val duration = effect.endFrame - startFrame + 1

        val progress = easingFunction.getProgress(effect.easing, currentFrame, duration)
        if (progress < 0 || progress > 1) {
            throw IllegalStateException("Animation progress is out of bounds: $progress")
        }

        return progress
    }

    protected abstract fun render(ctx: RenderContext)
}